src/test/miniminer_tests.cpp

```diff
@@ -0,0 +1,467 @@
+        BOOST_CHECK(!mini_miner_tx7_tx8.IsReadyToCalculate());
+        BOOST_CHECK(tx7_tx8_bumpfee.has_value());
+        BOOST_CHECK_EQUAL(tx7_tx8_bumpfee.value(), just_below_tx5.GetFee(tx_vsizes[5] + tx_vsizes[6]) - (low_fee + med_fee));
+    }
+    // Feerate between tx7 and tx8's ancestor feerates: don't need to bump tx6 because tx8 already does.
+    {
+        const auto just_above_tx7 = CFeeRate(med_fee + 10, tx_vsizes[6]);
+        BOOST_CHECK(just_above_tx7 <= CFeeRate(low_fee + high_fee, tx_vsizes[5] + tx_vsizes[7]));
+        node::MiniMiner mini_miner(pool, all_unspent_outpoints);
+        BOOST_CHECK(mini_miner.IsReadyToCalculate());
+        auto bump_fees = mini_miner.CalculateBumpFees(just_above_tx7);
+        BOOST_CHECK(!mini_miner.IsReadyToCalculate());
+        BOOST_CHECK_EQUAL(bump_fees.size(), all_unspent_outpoints.size());
+        BOOST_CHECK(sanity_check(all_transactions, bump_fees));
+        const auto tx7_bumpfee = bump_fees.find(COutPoint{tx7->GetHash(), 0});
+        BOOST_CHECK(tx7_bumpfee != bump_fees.end());
+        BOOST_CHECK_EQUAL(tx7_bumpfee->second, just_above_tx7.GetFee(tx_vsizes[6]) - (med_fee));
+        const auto tx8_bumpfee = bump_fees.find(COutPoint{tx8->GetHash(), 0});
+        BOOST_CHECK(tx8_bumpfee != bump_fees.end());
+        BOOST_CHECK_EQUAL(tx8_bumpfee->second, 0);
+    }
+}
+BOOST_FIXTURE_TEST_CASE(calculate_cluster, TestChain100Setup)
+{
+    FastRandomContext det_rand{true};
+    CTxMemPool& pool = *Assert(m_node.mempool);
+    LOCK2(cs_main, pool.cs);
+
+    // Add chain of size 500
+    TestMemPoolEntryHelper entry;
+    std::vector<uint256> chain_txids;
+    auto& lasttx = m_coinbase_txns[0];
+    for (auto i{0}; i < 500; ++i) {
+        const auto tx = make_tx({COutPoint{lasttx->GetHash(), 0}}, /*num_outputs=*/1);
+        pool.addUnchecked(entry.Fee(CENT).FromTx(tx));
+        chain_txids.push_back(tx->GetHash());
+        lasttx = tx;
+    }
+    const auto cluster_500tx = pool.GatherClusters({lasttx->GetHash()});
+    CTxMemPool::setEntries cluster_500tx_set{cluster_500tx.begin(), cluster_500tx.end()};
+    BOOST_CHECK_EQUAL(cluster_500tx.size(), cluster_500tx_set.size());
+    const auto vec_iters_500 = pool.GetIterVec(chain_txids);
+    for (const auto& iter : vec_iters_500) BOOST_CHECK(cluster_500tx_set.count(iter));
+
+    // GatherClusters stops at 500 transactions.
+    const auto tx_501 = make_tx({COutPoint{lasttx->GetHash(), 0}}, /*num_outputs=*/1);
+    pool.addUnchecked(entry.Fee(CENT).FromTx(tx_501));
+    const auto cluster_501 = pool.GatherClusters({tx_501->GetHash()});
+    BOOST_CHECK_EQUAL(cluster_501.size(), 0);
+
+    // Zig Zag cluster:
+    // txp0     txp1     txp2    ...  txp48  txp49
+    //    \    /    \   /   \            \   /
+    //     txc0     txc1    txc2  ...    txc48
+    // Note that each transaction's ancestor size is 2 or 3, and each descendant size is 2 or 3.
```

Comment:
pedantic-nit:
```suggestion
    // Note that each transaction's ancestor size is 1 or 3, and each descendant size is 1, 2 or 3.
```
The descendant count of 2 is the rare case here, only applying for parent-txs txp0 and txp49
(for the fun of it, created a functional test for verifying this: https://github.com/theStack/bitcoin/blob/functional_test_mempool_zigzag_playground/test/functional/mempool_zigzag.py).